File: app/schemas/schedule.py
"""Schedule schemas for time blocking."""
from typing import List, Optional
from datetime import datetime, date, time
from pydantic import BaseModel


class TimeBlock(BaseModel):
    """A time block for a task."""
    task_id: str
    title: str
    scheduled_time: datetime
    duration: int  # minutes
    status: str
    evidence_type: Optional[str] = None
    project_id: Optional[str] = None


class DueTask(BaseModel):
    """Task due on a specific date (deadline)."""
    task_id: str
    title: str
    status: str
    deadline: datetime
    priority: Optional[str] = None
    project_id: Optional[str] = None


class DailySchedule(BaseModel):
    """Daily schedule with time blocks and due tasks."""
    date: date
    time_blocks: List[TimeBlock]
    due_tasks: List[DueTask] = []
    
    class Config:
        json_schema_extra = {
            "example": {
                "date": "2026-02-02",
                "time_blocks": [
                    {
                        "task_id": "task-uuid",
                        "title": "起床",
                        "scheduled_time": "2026-02-02T07:00:00",
                        "duration": 30,
                        "status": "OPEN",
                        "evidence_type": "none"
                    }
                ]
            }
        }


class WeeklySchedule(BaseModel):
    """Weekly schedule with daily schedules."""
    start_date: date
    end_date: date
    daily_schedules: List[DailySchedule]


class ScheduleTaskRequest(BaseModel):
    """Request to schedule a task."""
    scheduled_date: date
    scheduled_time: time
    duration: int  # minutes


class ScheduleTaskResponse(BaseModel):
    """Response after scheduling a task."""
    task_id: str
    scheduled_time: datetime
    duration: int
    is_time_blocked: bool

File: frontend/lib/api/schedule.ts
// Schedule API client

const API_BASE = "";

export interface TimeBlock {
    task_id: string;
    title: string;
    scheduled_time: string;
    duration: number;
    status: string;
    evidence_type?: string;
    project_id?: string;
}

export interface DueTask {
    task_id: string;
    title: string;
    status: string;
    deadline: string;
    priority?: string;
    project_id?: string;
}

export interface DailySchedule {
    date: string;
    time_blocks: TimeBlock[];
    due_tasks: DueTask[];
}

export interface WeeklySchedule {
    start_date: string;
    end_date: string;
    daily_schedules: DailySchedule[];
}

export async function getTodaySchedule(token: string): Promise<DailySchedule> {
    const response = await fetch(`${API_BASE}/api/schedule/today`, {
        headers: {
            Authorization: `Bearer ${token}`,
        },
    });

    if (!response.ok) {
        throw new Error("Failed to fetch today's schedule");
    }

    return response.json();
}

export async function getWeekSchedule(token: string, startDate?: string): Promise<WeeklySchedule> {
    const query = startDate ? `?start_date=${startDate}` : "";
    const response = await fetch(`${API_BASE}/api/schedule/week${query}`, {
        headers: {
            Authorization: `Bearer ${token}`,
        },
    });

    if (!response.ok) {
        throw new Error("Failed to fetch week schedule");
    }

    return response.json();
}

export async function scheduleTask(
    taskId: string,
    scheduledDate: string,
    scheduledTime: string,
    duration: number,
    token: string
): Promise<any> {
    const response = await fetch(`${API_BASE}/api/schedule/tasks/${taskId}/schedule`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify({
            scheduled_date: scheduledDate,
            scheduled_time: scheduledTime,
            duration,
        }),
    });

    if (!response.ok) {
        throw new Error("Failed to schedule task");
    }

    return response.json();
}

export async function unscheduleTask(
    taskId: string,
    token: string
): Promise<void> {
    const response = await fetch(`${API_BASE}/api/schedule/tasks/${taskId}/schedule`, {
        method: "DELETE",
        headers: {
            Authorization: `Bearer ${token}`,
        },
    });

    if (!response.ok) {
        throw new Error("Failed to unschedule task");
    }
}
